ethereum/forge-test/Bridge.t.sol

```diff
@@ -56,6 +146,1289 @@ contract TestBridge is Test {
+            toChain,
+            addressToBytes32(address(this)),
+            additionalPayload
+        );
+        vm.expectEmit();
+        emit LogMessagePublished(
+            address(bridge),
+            uint64(1),
+            uint32(234),
+            transferPayload,
+            uint8(finality)
+        );
+        bridge.transferTokensWithPayload(
+            address(tokenImpl),
+            amount,
+            toChain,
+            toAddress,
+            234,
+            additionalPayload
+        );
+
+        uint256 accountBalanceAfter = tokenImpl.balanceOf(address(this));
+        uint256 bridgeBalanceAfter = tokenImpl.balanceOf(address(bridge));
+
+        assertEq(accountBalanceAfter, 0);
+        assertEq(bridgeBalanceAfter, amount);
+    }
+
+    function testShouldTransferOutLockedAssetsForAValidTransferWithPayloadVM()
+        public
+    {
+        testShouldDepositAndLogTransfersCorrectly();
+
+        uint256 amount = 1_000_000_000_000_000_000;
+        uint64 sequence = 1111;
+
+        uint256 accountBalanceBefore = tokenImpl.balanceOf(address(this));
+        uint256 bridgeBalanceBefore = tokenImpl.balanceOf(address(bridge));
+        assertEq(accountBalanceBefore, 0);
+        assertEq(bridgeBalanceBefore, amount);
+
+        bytes memory transferPayload = abi.encodePacked(
+            uint8(3),
+            amount / 1e10,
+            addressToBytes32(address(tokenImpl)),
+            testChainId,
+            addressToBytes32(address(this)),
+            testChainId,
+            uint256(0),
+            bytes("abc123")
+        );
+
+        bytes memory vaa = signAndEncodeVM(
+            0,
+            0,
+            testForeignChainId,
+            testForeignBridgeContract,
+            sequence,
+            transferPayload,
+            uint256Array(testGuardian),
+            0,
+            0
+        );
+
+        vm.expectEmit();
+        emit TransferRedeemed(
+            testForeignChainId,
+            testForeignBridgeContract,
+            sequence
+        );
+        bridge.completeTransferWithPayload(vaa);
+
+        uint256 accountBalanceAfter = tokenImpl.balanceOf(address(this));
+        uint256 bridgeBalanceAfter = tokenImpl.balanceOf(address(bridge));
+        assertEq(accountBalanceAfter, amount);
+        assertEq(bridgeBalanceAfter, 0);
+    }
+
+    function testShouldMintBridgedAssetWrappersOnTransferFromAnotherChainAndHandleFeesCorrectly()
+        public
+    {
+        testShouldTransferOutLockedAssetsForAValidTransferWithPayloadVM();
+
+        uint256 amount = 1_000_000_000_000_000_000;
+        uint256 fee = 100_000_000_000_000_000;
+
+        address wrappedAddress = bridge.wrappedAsset(
+            testBridgedAssetChain,
+            testBridgedAssetAddress
+        );
+        TokenImplementation wrapped = TokenImplementation(wrappedAddress);
+
+        assertEq(wrapped.totalSupply(), 0, "Wrong total supply");
+
+        bytes memory data = abi.encodePacked(
```

Comment:
Instead of formulating the wormhole payload in each test, there should be a helper method that creates a token bridge transfer payload.

Reply:
Agree this should be addressed with #4090